spring-cloud-function-integration/src/main/java/org/springframework/cloud/function/integration/dsl/FunctionFlowBuilder.java

```diff
@@ -0,0 +1,172 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.function.integration.dsl;
+
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import org.reactivestreams.Publisher;
+
+import org.springframework.cloud.function.context.FunctionCatalog;
+import org.springframework.integration.core.MessageSource;
+import org.springframework.integration.dsl.GatewayProxySpec;
+import org.springframework.integration.dsl.IntegrationFlow;
+import org.springframework.integration.dsl.IntegrationFlowBuilder;
+import org.springframework.integration.dsl.MessageChannelSpec;
+import org.springframework.integration.dsl.MessageProducerSpec;
+import org.springframework.integration.dsl.MessageSourceSpec;
+import org.springframework.integration.dsl.MessagingGatewaySpec;
+import org.springframework.integration.dsl.SourcePollingChannelAdapterSpec;
+import org.springframework.integration.endpoint.MessageProducerSupport;
+import org.springframework.integration.gateway.MessagingGatewaySupport;
+import org.springframework.lang.Nullable;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.util.Assert;
+
+/**
+ * The entry point for starting a {@link FunctionFlowDefinition}.
+ * Requires a {@link FunctionCatalog} to lookup function instances
+ * by their names or definitions from respective operators.
+ * <p>
+ * In addition to standard {@link IntegrationFlow} {@code from()} overloaded methods (for convenience),
+ * this class introduces {@link #fromSupplier(String)} factory methods to resolve the target {@link Supplier}
+ * by its name or function definition from the provided {@link FunctionCatalog}.
+ * <p>
+ * This class represents a DSL for functions composition via integration endpoints.
+ * Extra processing can be done in between functions by the regular {@link IntegrationFlow} operators:
+ * <pre class="code">
+ * {@code
+ * @Bean
+ * IntegrationFlow someFunctionFlow(FunctionFlowBuilder functionFlowBuilder) {
+ *		return functionFlowBuilder
+ *				.fromSupplier("timeSupplier")
+ *				.apply("spelFunction")
+ *				.log(LoggingHandler.Level.DEBUG, "some.log.category")
+ *				.<String, String>transform(String::toUpperCase)
+ *				.accept("fileConsumer");
+ * }
+ * }
+ * </pre>
+ *
+ * @author Artem Bilan
+ *
+ * @since 4.0.3
+ */
+public class FunctionFlowBuilder {
+
+	private final FunctionLookupHelper functionLookupHelper;
+
+	public FunctionFlowBuilder(FunctionCatalog functionCatalog) {
+		Assert.notNull(functionCatalog, "'functionCatalog' must not be null");
+		this.functionLookupHelper = new FunctionLookupHelper(functionCatalog);
```

Comment:
I don't think so. This is just a supporting internal class which does not have a `public` ctor.
Original I thought about just having it as a utility class, but then I decided to get rid of extra param for those `static` methods since this an internal stuff of this DSL logic.
Nothing to see here for end-user 😄

Reply:
If it is internal and not intended to be extended by user then - fair enough. 😸